### PR TITLE
Sets up the Python `gradio` client

### PR DESCRIPTION
**Pull request by abidlabs**
Creates a lightweight Python library called `gradio_client`

Documentation:

-----------

# `gradio_client`: Use any Gradio app as an API -- in 3 lines of Python

## Installation

The lightweight `gradio_client` package can be installed from `pip` (or `pip3`) and works with Python versions 3.7 or higher:

```bash
$ pip install gradio_client
```

### Connecting to a Space or a Gradio app

Start by connecting instantiating a `Client` object and connecting it to a Gradio app 
that is running on Spaces (or anywhere else)!

**Connecting to a Space**

```python
import gradio_client as grc

client = grc.Client(space="abidlabs/en2fr")
```

**Connecting a general Gradio app**

If your app is running somewhere else, provide the full URL instead to the `src` argument. Here's an example of making predictions to a Gradio app that is running on a share URL:

```python
import gradio_client as grc

client = grc.Client(src="btd372-js72hd.gradio.app")
```

### Making a prediction

The simplest way to make a prediction is simply to call the `.predict()` function with the appropriate arguments and then immediately calling `.result()`, like this:

```python
import gradio_client as grc

client = grc.Client(space="abidlabs/en2fr")

client.predict("Hello").result()

>> Bonjour
```

**Running jobs asyncronously**

Oe should note that `.result()` is a *blocking* operation as it waits for the operation to complete before returning the prediction. 

In many cases, you may be better off letting the job run asynchronously and waiting to call `.result()` when you need the results of the prediction. For example:

```python
import gradio_client as grc

client = grc.Client(space="abidlabs/en2fr")

job = client.predict("Hello")

# Do something else

job.result()

>> Bonjour
```

**Adding callbacks**

Alternatively, one can add callbacks to perform actions after the job has completed running, like this:

```python
import gradio_client as grc

def print_result(x):
    print(x"The translated result is: {x}")

client = grc.Client(space="abidlabs/en2fr")

job = client.predict("Hello", result_callbacks=[print_result])
```

Closes: #2516

### CODE REVIEW

**Comment by pngwn** [comment]
I've created #3310 and renamed #2516 so we can close them separately.

**Comment by freddyaboulton** [comment]
Very nice PR @abidlabs ! 🔥 

> By high-level, I mean referring to the organization of files and the usage syntax.

### Regarding organization of files
I think the addition of `client` directory with sub-directories for each language makes sense. Nothing much to say here from my end!

### Regarding usage
I like the way clients are instantiated and how the `predict` method is used. I think that's easy to use and intuitive. And the API introspection is really cool! Questions/Notes I had:

1. How does `predict` behave in the case of generator functions?
2. On a similar note, I think we should allow callbacks for each of the queue messages, not just when the job is complete. Some use cases I can see for this: providing ETA updates from the upstream queue, stitching together the intermediate outputs of a diffusion model into a gif. 
3. I see the api introspection only documents named routes. I think that's reasonable but I have a sense that most routes are actually unnamed. I wonder if we can still provide something for the un-named routes? We have all the input component info in the config even if the route is unnamed so it should still be possible to document them.

 Will think about this some more over the weekend!

**Comment by abidlabs** [comment]
Thanks for the feedback @freddyaboulton! 

> 1. How does predict behave in the case of generator functions?

Still need to handle this case. I'm planning on doing something similar to @pngwn's js client, where predict() gets you the first output of a generator

> 2. On a similar note, I think we should allow callbacks for each of the queue messages, not just when the job is complete. Some use cases I can see for this: providing ETA updates from the upstream queue, stitching together the intermediate outputs of a diffusion model into a gif.

Sounds good. If it's straightforward enough, I'll add to this PR, otherwise will save for a future PR

> 3. I see the api introspection only documents named routes. I think that's reasonable but I have a sense that most routes are actually unnamed. I wonder if we can still provide something for the un-named routes? We have all the input component info in the config even if the route is unnamed so it should still be possible to document them.

I was thinking exactly the same thing. For Blocks, routes are unnamed by default so many of the popular Spaces don't show any named routes right now. I don't have a good solution for this because, although we can certainly provide the information for unnamed routes, this would include *every* single event listener in the app which could be a lot. Suggestions welcome! (cc @aliabid94)

File: client/python/gradio_client/client.py
Comment (Wauplin) [inline client/python/gradio_client/client.py:28]: Would be a could use case for `huggingface_hub` here:

```py
from huggingface_hub.utils import build_hf_headers

headers = build_hf_headers(token=access_token)
# or even
headers = build_hf_headers(token=access_token, library_name="gradio_client", library_version"0.xx")
```
Especially helpful if the user has used `huggingface-cli login` or `huggingface_hub.login` (+it adds a user-agent if you are interested in telemetry)

File: client/python/gradio_client/client.py
Comment (Wauplin) [inline client/python/gradio_client/client.py:34]: It seems `self.space` is never used except in `__init__.py` to compute the src url.
I would rather not set it and pass an arg to `_space_name_to_src`.

File: client/python/gradio_client/client.py
Comment (Wauplin) [inline client/python/gradio_client/client.py:116]: (nit) it feels more readable to me to condense it like this:

```suggestion
        for i, d in enumerate(self.config["dependencies"]):
            if d.get("api_name") == api_name:
                return i
        raise ValueError(f"Cannot find a function with api_name: {api_name}")
```

File: client/python/gradio_client/utils.py
Comment (Wauplin) [inline client/python/gradio_client/utils.py:76]: Could use `huggingface_hub.build_hf_headers` here as well

File: client/python/test/test_utils.py
Comment (Wauplin) [inline client/python/test/test_utils.py:11]: unused?
```suggestion
```

**Review by Wauplin — COMMENTED** [review]
Nice python client @abidlabs! I started to read it and ended up with a few very low-level comments -you can discard them if you want-. My main feedback is that I think we can use `huggingface_hub.build_hf_headers` to handle authentication. It would be more Hub-focused rather than generic to Gradio but it would ease things for users that are already used to login via the cli/in a notebook.

**Comment by abidlabs** [comment]
Thanks @Wauplin! Appreciate the suggestions, will utilize `huggingface_hub` more

**Comment by abidlabs** [comment]
Question for you: is there a way to replace: 

```py
    def _space_name_to_src(self) -> str | None:
        return (
            requests.get(
                f"https://huggingface.co/api/spaces/{self.space}/host",
                headers=self.headers,
            )
            .json()
            .get("host")
        )
```

with a function from `huggingface_hub`? I think when we discussed it last, there wasn't a way but I wanted to double check

**Comment by Wauplin** [comment]
> Question for you Wauplin: is there a way to replace `_space_name_to_src` with a function from huggingface_hub?

This is very new from the Hub side but yes it is possible:

```py
from huggingface_hub import space_info

def _space_name_to_src(self) -> str | None:
    info = space_info("microsoft/visual_chatgpt")
    return f"https://{info.subdomain}.hf.space"
```

It seems that `host` is not returned but `subdomain` is. I am not sure API-side is you are guaranteed to have a subdomain so you might need to to `subdomain = getattr(info, "subdomain", None); if subdomain is not None: ...`.

(cc @coyotte508 @julien-c since I know returning "host" has been discussed _at some point_ but couldn't retrieve the convo)

**Comment by coyotte508** [comment]
You are guaranteed to get a `subdomain`.

Regarding host, it'd be nice to have I agree

**Comment by julien-c** [comment]
FWIW there's also a `GET /api/:repoType(spaces)/:namespace/:repo/host` endpoint which is already used from the gradio codebase

File: client/python/gradio_client/client.py
Comment (abidlabs) [reply client/python/gradio_client/client.py]: Done

**Comment by abidlabs** [comment]
Hi everyone! This PR is ready for review. It's not quite ready for launch yet (and I think we should hold off from launching until after theming), so I've removed it from the changelog for now. But the core pieces are in place and I did want to merge it in before it becomes stale, since a lot of files are affected. Specifically, this PR:

* Creates a `client/python/gradio_client` directory for the new `gradio_client` library
* Adds various scripts / tests / packaging-related files so that we can easily test the `gradio_client` and release it to pypi as needed
* Moves various pieces of the `gradio` library to `gradio_client`, most importantly `serializing.py`
* Of course, creates the core `gradio_client.Client` class itself. 
* Ensures all old tests continue to pass while adding a few new ones

Here are the things that are still left to do (I'll create a tracking issue afterwards):

* add support for iterators -- right now iterators throw return a meaningless update as a result
* add status callbacks, not just result callbacks
* add ability to inspect and quick get an idea of how to use each of the API endpoints for a given gradio app
* add this PR back to the changelog 

Please let me know if you find any other gaps or bugs as you test it out. Instructions to test are at the top of the PR.

**Comment by abidlabs** [comment]
From our sync today, adding a few more TODOs:

- Update this PR to resolve issues related to temp files
- Document gr.load only
- rename acces_token to hf_token
- clarify that callbacks are executed sequentially
- Files should be uploaded via /upload
- Session management / reset()
- Analytics 
- HuggingFace Hub as a dependency

Some of these will be tackled in future PRs

**Comment by abidlabs** [comment]
Made some changes to this PR:

* Reconciled with `main` 
* Documented `gr.load()`. I included the documentation for `gr.Blocks.load()` and `gr.Interface.load()` but marked them as methods that will be deprecated:

image


image

* Added telemetry using the `huggingface_hub` library

File: client/python/README.md
Comment (aliabid94) [inline client/python/README.md:3]: can we remove this comment? I think it discourages users unnecessarily - if this is trusted enough to be used for the core gradio library, then its good enough for our users, they'll open issues if things break

File: client/python/README.md
Comment (aliabid94) [inline client/python/README.md:16]: unclear `>> URL` was supposed to be the output from `job.result()`

File: client/python/gradio_client/client.py
Comment (aliabid94) [inline client/python/gradio_client/client.py:66]: allow result_callbacks to be a singleton, and wrap as list if so?

File: client/python/gradio_client/client.py
Comment (aliabid94) [inline client/python/gradio_client/client.py:92]: dont like this big comment block, private methods are self evident via underscore prefix

**Review by aliabid94 — APPROVED** [review]
lgtm!

File: client/python/README.md
Comment (abidlabs) [reply client/python/README.md:3]: I'll soften the message, but the library is not meant for external release. We can always change the names of variables, etc. for internal use but if people build on top of it, then it may cause their code to break.

File: client/python/README.md
Comment (abidlabs) [reply client/python/README.md:16]: Ok I'll just remove this

File: client/python/README.md
Comment (abidlabs) [reply client/python/README.md:104]: Ok I prefer `callbacks` too but we are also planning on adding `status_callbacks`

Comment (abidlabs) [reply client/python/README.md:104]: We can change this later if needed (since library is in alpha)

File: client/python/gradio_client/client.py
Comment (abidlabs) [reply client/python/gradio_client/client.py:66]: Sure sg

File: client/python/gradio_client/client.py
Comment (abidlabs) [reply client/python/gradio_client/client.py:92]: Sg

File: client/python/gradio_client/client.py
Comment (freddyaboulton) [inline client/python/gradio_client/client.py:184]: Right now there is no way to use a space with auth via api right? I feel like that is something we should support? Curious on your thoughts

Comment (freddyaboulton) [reply client/python/gradio_client/client.py:184]: Maybe we should set is_valid to false for endpoints in auth enabled apps?

File: client/python/gradio_client/client.py
Comment (freddyaboulton) [inline client/python/gradio_client/client.py:196]: Curious how come we're returning tuples everywhere given that "data" is already a list?

File: gradio/external.py
Comment (freddyaboulton) [inline gradio/external.py:34]: nit: I think the way `set_documentation_droup` works, it doesn't really allow calling it twice because it resets the classes to document for that group every time it's called. 

Just tested locally and this is not showing up. 

Let's fix before next release!

File: scripts/format_backend.sh
Comment (freddyaboulton) [inline scripts/format_backend.sh:8]: I think we need an `--exclude` for each file?

This script isn't working for me

image

File: test/test_blocks.py
Comment (freddyaboulton) [inline test/test_blocks.py:129]: It's in a test file but let's still get rid of it

**Review by freddyaboulton — COMMENTED** [review]
Super awesome PR @abidlabs !

Left some comments that can be handled in future PRs

**Comment by abidlabs** [comment]
Thanks for the review @freddyaboulton and @aliabid94! Will address these comments

File: client/python/scripts/ci.sh
Comment (abidlabs) [reply client/python/scripts/ci.sh:8]: Fixed thanks

File: client/python/gradio_client/client.py
Comment (abidlabs) [reply client/python/gradio_client/client.py:120]: Done

File: client/python/gradio_client/client.py
Comment (abidlabs) [reply client/python/gradio_client/client.py:184]: Let me add it as a list of things to support in the tracking issue

File: client/python/gradio_client/client.py
Comment (abidlabs) [reply client/python/gradio_client/client.py:196]: It helped with the typing I believe

**Comment by abidlabs** [comment]
Ok I think that should be everything! Will merge in after CI passes

File: gradio/external.py
Comment (freddyaboulton) [reply gradio/external.py:34]: Not the first time that has tripped me up 🤦 Disregard!

File: scripts/format_backend.sh
Comment (abidlabs) [reply scripts/format_backend.sh]: Fixed. Turns out you can use comma (no space) to separate multiple filenames

**Pull request merged**
